### DISCUSSION TITLE
if then else - syntax sugar for control flow with the <Show> component

### DISCUSSION THREAD

**Discussion by milahu** [solidjs/solid#1221] [Q&A] [answered]
[control flow](https://www.solidjs.com/tutorial/flow_show) is one of my pain points with solidjs

_i assume this was discussed before, but i cant find a thread or issue_

```jsx
export default function App(props) {
  return (
    <Show
      when={props.condition}
      fallback={
        <Show
          when={props.condition2}
          fallback={<>alternative</>}
        >expression2</Show>
      }
    >expression</Show>
  );
}
```

<details>
<summary>
full code
</summary>

```tsx
import { render } from "solid-js/web";
import { Show, createSignal } from "solid-js";

export default function App(props) {
  const [getCondition, setCondition] = createSignal(false);
  const [getCondition2, setCondition2] = createSignal(false);
  return (
    <div>
      <ul>
        <li>
          <label>
            <input type="checkbox" checked={getCondition()} onChange={(event) => (console.log(event.target.checked), setCondition(event.target.checked))} />
            condition
          </label>
        </li>
        <li>
          <label>
            <input type="checkbox" checked={getCondition2()} onChange={(event) => (console.log(event.target.checked), setCondition2(event.target.checked))} />
            condition2
          </label>
        </li>
      </ul>
      <Show
        when={getCondition()}
        fallback={
          <Show
          when={getCondition2()}
            fallback={<>alternative</>}
          >expression2</Show>
        }
      >expression</Show>
    </div>
  );
}

render(() => <App/>, document.getElementById("app")!);
```

</details>

compare this with svelte

```svelte
<script>
  export let condition;
  export let condition2;
</script>

{#if condition}
  expression
{#elif condition2}
  expression2
{#else}
  alternative
{#/if}
```

i assume this is bad for performance

```jsx
export default function App(props) {
  return (
    <If cond={props.condition}>
      <Then>expression</Then>
      <Elif cond={props.condition2}>expression2</Elif>
      <Else>alternative</Else>
    </If>
  );
}
```

... but why?

cant we have something like a "tail call optimization" for solidjs
which would reduce the 4 components `If` `Then` `Elif` `Else` into one component?

or implement this with desugaring in a solidjs preprocessor plugin ...

**Comment by ryansolid** [comment, marked as answer]
Have you seen Switch/Match? https://www.solidjs.com/tutorial/flow_switch

There'd be no perf issue with your last example. We do everything lazy and don't really have components. And JSX can return anything. It just felt awkward syntactically. I looked at a number of these syntaxes.

We really don't want to use a pre-processor in core for components. Solid's components just being runtime components is very important. In fact, if you like what you wrote you can optimally make it without a compiler. That's the power here. Look at the implementation of Switch/Match for hints how to. https://github.com/solidjs/solid/blob/main/packages/solid/src/render/flow.ts#L120-L184

**Reply by milahu** [reply]
`<Switch>` is almost perfect ...

the fallback branch should come last in the JSX code
which currently is possible with `<Match when={true}>`

to make this more pretty,
the default value for `<Match>.props.when` should be `true`

```jsx
function App() {
  const [x] = createSignal(7);

  return (
    <Switch
      // fallback={<p>{x()} is between 5 and 10</p> // this is ugly
    >
      <Match when={x() > 10} >
        <p>{x()} is greater than 10</p>
      </Match>
      <Match when={5 > x()}>
        <p>{x()} is less than 5</p>
      </Match>
      <Match>
        <p>{x()} is between 5 and 10 (expected branch)</p>
      </Match>
      <Match when={true}>
        <p>{x()} is between 5 and 10 (actual branch)</p>
      </Match>
    </Switch>
  );
}
```

edit:

`<Switch>` should be mentioned in https://www.solidjs.com/tutorial/flow_show

edit:

looking at `solid/src/render/flow.ts`
a "default true" version is probably slower

```diff
         for (let i = 0; i < conds.length; i++) {
           const c = conds[i].when;
-          if (c) {
+          if (c === undefined || c) {
             keyed = !!conds[i].keyed;
             return [i, c, conds[i]];
           }
         }
```
